src/workerd/api/queue.c++

```diff
@@ -544,76 +560,124 @@ kj::Promise<WorkerInterface::CustomEvent::Result> QueueCustomEventImpl::run(
+    auto startResp = startQueueEvent(lock.getGlobalScope(), kj::mv(params),
         context.addObject(result), lock,
         lock.getExportedHandler(entrypointName, kj::mv(props), context.getActor()), typeHandler);
-  }));
-
-  // TODO(soon): There's a good chance we'll want a different wall-clock timeout for queue handlers
-  // than for scheduled workers, but it's not at all clear yet to me what it should be, so just
-  // reuse the scheduled worker logic and timeout for now.
-  auto result = co_await incomingRequest->finishScheduled();
-  bool completed = result == IoContext_IncomingRequest::FinishScheduledResult::COMPLETED;
-
-  // Log some debug info if the request timed out or was aborted.
-  // In particular, detect whether or not the users queue() handler function completed
-  // and include info about other waitUntil tasks that may have caused the request to timeout.
-  if (!completed) {
-    kj::String status;
-    if (queueEventHolder->event.get() == nullptr) {
-      status = kj::str("Empty");
-    } else {
-      KJ_SWITCH_ONEOF(queueEventHolder->event->getCompletionStatus()) {
-        KJ_CASE_ONEOF(i, QueueEvent::Incomplete) {
-          status = kj::str("Incomplete");
-          break;
-        }
-        KJ_CASE_ONEOF(s, QueueEvent::CompletedSuccessfully) {
-          status = kj::str("Completed Succesfully");
-          break;
-        }
-        KJ_CASE_ONEOF(e, QueueEvent::CompletedWithError) {
-          status = kj::str("Completed with error:", e.error);
-          break;
-        }
+    queueEvent->event = kj::mv(startResp.event);
+    queueEvent->exportedHandlerProm = kj::mv(startResp.exportedHandlerProm);
+    queueEvent->isServiceWorkerHandler = startResp.isServiceWorkerHandler;
+  });
+
+  auto compatFlags = context.getWorker().getIsolate().getApi().getFeatureFlags();
+  if (compatFlags.getQueueConsumerNoWaitForWaitUntil()) {
+    // The user has opted in to only waiting on their event handler rather than all waitUntil'd
+    // promises.
+    auto timeoutPromise = context.getLimitEnforcer().limitScheduled();
+    auto outcome = co_await runProm
+                       .then([queueEvent = kj::addRef(
+                                  *queueEventHolder)]() mutable -> kj::Promise<EventOutcome> {
+      KJ_IF_SOME(handlerProm, queueEvent->exportedHandlerProm) {
+        return handlerProm.then([]() { return EventOutcome::OK; });
       }
+      return EventOutcome::OK;
+    }).catch_([](kj::Exception&& e) {
+      return EventOutcome::EXCEPTION;
+    }).exclusiveJoin(timeoutPromise.then([] {
+      return EventOutcome::EXCEEDED_CPU;
+    })).exclusiveJoin(context.onAbort().then([] {
+      // This is a change from the outcome we returned on abort before the compat flag, but better
+      // matches the behavior of fetch() handlers and the semantics of what's actually happening.
+      return EventOutcome::EXCEPTION;
+    }, [](kj::Exception&&) { return EventOutcome::EXCEPTION; }));
```

Comment:
nit... just noting, following these promise chains melts the brain a bit.  A comment at the top explaining the flow here a bit more would be helpful for future newcomers looking into this code.

Reply:
Yeah, very fair. Added some comments to try to break it down.